builtin/add.c

```diff
@@ -208,7 +208,11 @@ static int refresh(int verbose, const struct pathspec *pathspec)
 		}
 	}
 
-	if (only_match_skip_worktree.nr) {
+	/*
+	 * When using a virtual filesystem, we might re-add a path
+	 * that is currently virtual and we want that to succeed.
+	 */
+	if (only_match_skip_worktree.nr && !core_virtualfilesystem) {
```

Comment:
How about leading with the `!core_virtualfilesystem` check? It is the less complicated and more obvious condition, therefore it removes cognitive load to "get it out of the way early".